README.md
<!---
  Licensed to the Apache Software Foundation (ASF) under one or more
  contributor license agreements.  See the NOTICE file distributed with
  this work for additional information regarding copyright ownership.
  The ASF licenses this file to You under the Apache License, Version 2.0
  (the "License"); you may not use this file except in compliance with
  the License.  You may obtain a copy of the License at

       http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
--->
:warning: Please note that this component is currently undergoing migration under the ASF PDFBox umbrella. The new Maven coordinates are not yet working. In the mean time please refer to the [latest released version](https://github.com/levigo/jbig2-imageio/tree/levigo-jbig2-imageio-2.0).

# Welcome to the Java ImageIO plugin for the JBIG2 image format!
The Java ImageIO plugin for JBIG2 enables access to images encoded using the JBIG2 image compression standard.
This component is part of the [Apache PDFBox® project](https://pdfbox.apache.org/) 

## Status
[![Build Status](https://travis-ci.org/levigo/jbig2-imageio.svg?branch=master)](https://travis-ci.org/levigo/jbig2-imageio)

## Features
The key features of the plugin are:

- Read-only (decode) support for images encoded using the JBIG2 standard, aka *ITU T.88* and *ISO/IEC 14492*.
- Full support for all features defined in the standard, including arithmetic coding, Huffman coding and MQ coding.
- Pure Java code. No funny native or JNI stuff required.
- Secure.
- Production quality. Robust and well-tested.
- Reasonable performance.
- Standard ImageIO API.
- Support for shared data segments as used by PDF via custom DecodeParams.

## License
This ImageIO plugin is licensed under the [Apache Software License 2.0](https://www.apache.org/licenses/LICENSE-2.0). Alternatively see [here](LICENSE.txt).

## Release Notes
Please take a look at the [release notes](release-notes.md).

## Support
Support is available through the [Apache PDFBox community}(https://pdfbox.apache.org/support.html)

## Usage
### Introduction
Please choose the appropriate section below, depending on whether you need to embed the ImageIO decoder inside your application (which would, for example, be the case if you want to support shared data segments) or you just want to add JBIG2 support to an existing ImageIO-enabled application.

### Use within existing Image I/O-based applications
Using the JBIG2 plugin with an existing application that already supports Java ImageIO is - at least in theory - very simple: just plunk the plugin jar down into your classpath and you're ready to go. What, exactly, you have to do to achieve this, depends on the application in question. Unfortunately, there is no general way to add the plugin to an application. Some general recommendations:

- If you implemented the application yourself: hey, don't look at us. You'll know best what to do. If you build your application using Maven, just pull in the plugin as a dependency. See below.
- Consult the application's manual (well, duh!).
- Maybe the application installation has a folder called 'plugins' or something like that. Try putting the jar into it. With some luck, the application will pick it up.
- Is the application started using a Unix shell script or Windows batch file? Try to identify where the classpath is assembled and add the plugin jar to it.
- As a last resort, you could try to add the plugin jar to the [lib/ext directory of your JRE installation](http://download.oracle.com/javase/1.4.2/docs/guide/extensions/spec.html). But please be advised that this is not considered to be good style.

### Use dependency management
The JBIG2 ImageIO-Plugin is available from [Maven Central](http://search.maven.org/). 

#### Maven
To use the plugin within a Maven POM-based project, simply include a dependency to the following artifact in the appropriate ```pom.xml```:

    <dependency>
      <groupId>org.apache.pdfbox.jbig2</groupId>
		<artifactId>pdfbox-jbig2-imageio</artifactId>
      <version>3.0-SNAPSHOT</version>
    </dependency>

### How to deal with embedded JBIG2 data
Several formats allow to embed JBIG2-compressed data in its own structure. PDF, for example, supports JBIG2-compressed data and adds the ability to embed shared data segments. Therefore the `JBIG2ImageReader` can handle `JBIG2Globals`  which are stored separately and can be passed into the reader if neccessary via `setGlobals()`-method.

The reader recognizes if the input data is headless and so embedded. This is assumed if the file header is missing.

You can also specify that the coming input data is embedded by using the special constructor in `JBIG2ImageReader`.

### What if the plugin is on classpath but not seen?
ImageIO is able to scan the classpath for readers and writers. Call `ImageIO.scanForPlugins()` if the reader is not seen. (Note: Thanks to George Sexton for this tip in context of using ImageIO within Apache Tomcat)


release-notes.md
<!---
  Licensed to the Apache Software Foundation (ASF) under one or more
  contributor license agreements.  See the NOTICE file distributed with
  this work for additional information regarding copyright ownership.
  The ASF licenses this file to You under the Apache License, Version 2.0
  (the "License"); you may not use this file except in compliance with
  the License.  You may obtain a copy of the License at

       http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
--->
# Release notes

## Version 3.0.0 (pending)
- [Issue #30](https://github.com/levigo/jbig2-imageio/issues/30): Transition project to Apache PDFBox.
- [Issue #25](https://github.com/levigo/jbig2-imageio/issues/25): Build tests fail if project path has a space.
- [Issue #26](https://github.com/levigo/jbig2-imageio/issues/26): Huffman user tables in text regions.
- [Issue #27](https://github.com/levigo/jbig2-imageio/issues/27): Problems in standard Huffman tables.
- [Issue #32](https://github.com/levigo/jbig2-imageio/issues/32): Newlines printed to stdout.

## Version 2.0.0 (2017-06-02)
- Java 9 support.
- [Issue #10](https://github.com/levigo/jbig2-imageio/issues/10): Use general service loading instead of Image I/O specific. See [merge request #18](https://github.com/levigo/jbig2-imageio/pull/18).
- [Issue #11](https://github.com/levigo/jbig2-imageio/issues/11): Remove @author info from javadocs.
- [Issue #16](https://github.com/levigo/jbig2-imageio/issues/16): Replace deprecated ```STANDARD_INPUT_TYPE``` constant with self-created array.
- [Issue #21](https://github.com/levigo/jbig2-imageio/issues/21): Move repetition of previous code length into correct scope. See [merge request #22](https://github.com/levigo/jbig2-imageio/pull/22). 

## Version 1.6.5 (2015-12-29)
- Issue #5: Deployment to maven central

## Version 1.6.3 (2014-08-04)
- Issue #1: Discard the request of copying the row above the first row.

## Version 1.6.2-RC1 (2014-01-22)
- Googlecode Issue 17: Changed computation of result dimension from ceiling to rounding. 

##Version 1.6.1 (2013-04-18)
- Googlecode Issue 11: Added support for `GRREFERENCE` Release notes

## Version 1.6.2-RC1 (2014-01-22)
- Googlecode Issue 17: Changed computation of result dimension from ceiling to rounding. 

## Version 1.6.1 (2013-04-18)
- Googlecode Issue 11: Added support for `GRREFERENCEDX` in generic refinement region coding. 

## Version 1.6.0 (2013-03-25):
- Googlecode Issue 10: Usability of `CacheFactory` and `LoggerFactory` improved.

## Version 1.5.2 (2012-10-09):
- Googlecode Issue  9: Transfer of bitmap's data into target raster went wrong if bitmap's line ends with a padded byte. 

## Version 1.5.1 (2012-10-02):
- Googlecode Issue  8: The default read parameters changed. There will be no source region, no source render size (no scaling) 
  and subsampling factors of 1 (no subsampling). `Bitmaps.java` can handle this correctly.

## Version 1.5.0 (2012-09-20):
- Moved Exception-classes to dedicated package `com.levigo.jbig2.err`.
- Introduced a new utility class `com.levigo.jbig2.image.Bitmaps`. This class provides a bunch of new features operating 
  on a `Bitmap` instance. For example, extracting a region of interest, scaling with high-quality filters and subsampling
  in either or both horizontal and vertical direction.

## Version 1.4.1 (2012-04-20):
- Fixed race condition when parsing documents with multiple pages.

## Version 1.4 (2012-04-10):
- Googlecode Issue  6 : The returned bitmap was too small in case of only one region. Solution is to check if we have only one
  region that forms the complete page. Only if width and height of region equals width and height of page use region's
  bitmap as the page's bitmap. 
- Googlecode Issue  5: A raster which was too small was created. AWT has thrown a `RasterFormatException`.  
- Googlecode Issue  4: `IndexOutOfBoundsException` indicates the end of the stream in `JBIG2Document#reachedEndOfStream()`
- Googlecode Issue  3: Reader recognizes if a file header is present or not.

## Version 1.3 (2011-10-28):
- Untracked Googlecode Issue : Fixed inverted color model for grayscale images.
- Untracked Googlecode Issue : Fixed `IndexArrayOutOfBoundException` in handling requests with region of interests. The region of
  interest is clipped at image boundary.

## Version 1.2 (2011-10-06):
- Googlecode Issue  1: The default read parameters will return a default image size of 1x1 without claiming the missing input.
- Untracked Googlecode Issue : A black pixel was represented by 1 and a white pixel by 0. For work with image masks the
  convention says, a black pixel is the minimum and the white pixel is maximum. This corresponds to an additive
  colorspace. We turned the representation of white and black pixels for conformity.

## Version 1.1 (2010-12-13):
- raster creation optimized
- potential NPE in cache 

## Version 1.0 (2010-07-29):
- open-source'd  DX in generic refinement region coding. 


src/main/java/com/levigo/jbig2/image/Filter.java
/**
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.levigo.jbig2.image;


abstract class Filter {

  /**
   * Find a filter name by its type.
   * 
   * @param type the filter type
   * @return filter name
   */
  public static String nameByType(final FilterType type) {
    if (type == null)
      throw new IllegalArgumentException("type must not be null");
    return type.name();
  }

  /**
   * Find a filter type by its name.
   * 
   * @param name the filter name
   * @return filter type
   */
  public static FilterType typeByName(final String name) {
    if (name == null)
      throw new IllegalArgumentException("name must not be null");
    return FilterType.valueOf(name);
  }

  /**
   * Find a filter by its type.
   * 
   * @param type the filter type
   * @return the Filter
   */
  public static Filter byType(final FilterType type) {
    switch (type){
      case Bessel :
        return new Bessel();
      case Blackman :
        return new Blackman();
      case Box :
        return new Box();
      case Catrom :
        return new Catrom();
      case Cubic :
        return new Cubic();
      case Gaussian :
        return new Gaussian();
      case Hamming :
        return new Hamming();
      case Hanning :
        return new Hanning();
      case Hermite :
        return new Hermite();
      case Lanczos :
        return new Lanczos();
      case Mitchell :
        return new Mitchell();
      case Point :
        return new Point();
      case Quadratic :
        return new Quadratic();
      case Sinc :
        return new Sinc();
      case Triangle :
        return new Triangle();
    }
    throw new IllegalArgumentException("No filter for given type.");
  }

  public static final class Bessel extends Filter {
    public Bessel() {
      super(false, 3.2383, 1.0);
    }

    private double J1(final double x) {
      double p, q;

      int i;

      final double Pone[] = {
          0.581199354001606143928050809e+21, -0.6672106568924916298020941484e+20, 0.2316433580634002297931815435e+19,
          -0.3588817569910106050743641413e+17, 0.2908795263834775409737601689e+15, -0.1322983480332126453125473247e+13,
          0.3413234182301700539091292655e+10, -0.4695753530642995859767162166e+7, 0.270112271089232341485679099e+4
      }, Qone[] = {
          0.11623987080032122878585294e+22, 0.1185770712190320999837113348e+20, 0.6092061398917521746105196863e+17,
          0.2081661221307607351240184229e+15, 0.5243710262167649715406728642e+12, 0.1013863514358673989967045588e+10,
          0.1501793594998585505921097578e+7, 0.1606931573481487801970916749e+4, 0.1e+1
      };

      p = Pone[8];
      q = Qone[8];
      for (i = 7; i >= 0; i--) {
        p = p * x * x + Pone[i];
        q = q * x * x + Qone[i];
      }
      return p / q;
    }

    private double P1(final double x) {
      double p, q;

      int i;

      final double Pone[] = {
          0.352246649133679798341724373e+5, 0.62758845247161281269005675e+5, 0.313539631109159574238669888e+5,
          0.49854832060594338434500455e+4, 0.2111529182853962382105718e+3, 0.12571716929145341558495e+1
      }, Qone[] = {
          0.352246649133679798068390431e+5, 0.626943469593560511888833731e+5, 0.312404063819041039923015703e+5,
          0.4930396490181088979386097e+4, 0.2030775189134759322293574e+3, 0.1e+1
      };

      p = Pone[5];
      q = Qone[5];
      for (i = 4; i >= 0; i--) {
        p = p * (8.0 / x) * (8.0 / x) + Pone[i];
        q = q * (8.0 / x) * (8.0 / x) + Qone[i];
      }
      return p / q;
    }

    private double Q1(final double x) {
      double p, q;

      int i;

      final double Pone[] = {
          0.3511751914303552822533318e+3, 0.7210391804904475039280863e+3, 0.4259873011654442389886993e+3,
          0.831898957673850827325226e+2, 0.45681716295512267064405e+1, 0.3532840052740123642735e-1
      }, Qone[] = {
          0.74917374171809127714519505e+4, 0.154141773392650970499848051e+5, 0.91522317015169922705904727e+4,
          0.18111867005523513506724158e+4, 0.1038187585462133728776636e+3, 0.1e+1
      };

      p = Pone[5];
      q = Qone[5];
      for (i = 4; i >= 0; i--) {
        p = p * (8.0 / x) * (8.0 / x) + Pone[i];
        q = q * (8.0 / x) * (8.0 / x) + Qone[i];
      }
      return p / q;
    }

    private double BesselOrderOne(double x) {
      double p, q;

      if (x == 0.0)
        return 0.0;
      p = x;
      if (x < 0.0)
        x = -x;
      if (x < 8.0)
        return p * J1(x);
      q = Math.sqrt(2.0 / (Math.PI * x))
          * (P1(x) * (1.0 / Math.sqrt(2.0) * (Math.sin(x) - Math.cos(x))) - 8.0 / x * Q1(x)
              * (-1.0 / Math.sqrt(2.0) * (Math.sin(x) + Math.cos(x))));
      if (p < 0.0)
        q = -q;
      return q;
    }

    @Override
    public double f(final double x) {
      if (x == 0.0)
        return Math.PI / 4.0;
      return BesselOrderOne(Math.PI * x) / (2.0 * x);
    }
  }

  public static final class Blackman extends Filter {
    @Override
    public double f(final double x) {
      return 0.42 + 0.50 * Math.cos(Math.PI * x) + 0.08 * Math.cos(2.0 * Math.PI * x);
    }
  }

  public static class Box extends Filter {
    public Box() {
      super(true, .5, 1.0);
    }

    public Box(final double supp) {
      super(true, supp, 1.0);
    }

    @Override
    public double f(final double x) {
      if (x >= -0.5 && x < 0.5)
        return 1.0;
      return 0.0;
    }
  }

  public static final class Point extends Box {
    public Point() {
      super(0);
    }

    @Override
    public double fWindowed(double x) {
      // don't apply windowing as we have a radius of zero.
      return super.f(x);
    }
  }

  public static final class Catrom extends Filter {
    public Catrom() {
      super(true, 2.0, 1.0);
    }

    @Override
    public double f(double x) {
      if (x < 0)
        x = -x;
      if (x < 1.0)
        return 0.5 * (2.0 + x * x * (-5.0 + x * 3.0));
      if (x < 2.0)
        return 0.5 * (4.0 + x * (-8.0 + x * (5.0 - x)));
      return 0.0;
    }
  }

  public static final class Cubic extends Filter {
    public Cubic() {
      super(false, 2.0, 1.0);
    }

    @Override
    public double f(double x) {
      if (x < 0)
        x = -x;
      if (x < 1.0)
        return 0.5 * x * x * x - x * x + 2.0 / 3.0;
      if (x < 2.0) {
        x = 2.0 - x;
        return 1.0 / 6.0 * x * x * x;
      }
      return 0.0;
    }
  }

  public static final class Gaussian extends Filter {
    public Gaussian() {
      super(false, 1.25, 1.0);
    }

    @Override
    public double f(final double x) {
      return Math.exp(-2.0 * x * x) * Math.sqrt(2.0 / Math.PI);
    }
  }

  public static final class Hamming extends Filter {
    @Override
    public double f(final double x) {
      return 0.54 + 0.46 * Math.cos(Math.PI * x);
    }
  }

  public static final class Hanning extends Filter {
    @Override
    public double f(final double x) {
      return 0.5 + 0.5 * Math.cos(Math.PI * x);
    }
  }

  public static final class Hermite extends Filter {
    @Override
    public double f(double x) {
      if (x < 0) {
        x = -x;
      }

      if (x < 1.0) {
        return (2.0 * x - 3.0) * x * x + 1.0;
      }
      return 0.0;
    }
  }

  public static final class Lanczos extends Filter {
    public Lanczos() {
      super(true, 3.0, 1.0);
    }

    @Override
    public double f(double x) {
      if (x < 0)
        x = -x;
      if (x < 3.0)
        return (float) (sinc(x) * sinc(x / 3.0));
      return 0.0;
    }

    private double sinc(double value) {
      if (value != 0.0f) {
        value = value * Math.PI;
        return Math.sin(value) / value;
      } else {
        return 1.0;
      }
    }

  }

  public static final class Mitchell extends Filter {
    public Mitchell() {
      super(false, 2.0, 1.0);
    }

    @Override
    public double f(double x) {
      double b, c;

      b = 1.0 / 3.0;
      c = 1.0 / 3.0;
      if (x < 0)
        x = -x;
      if (x < 1.0) {
        x = (12.0 - 9.0 * b - 6.0 * c) * (x * x * x) + (-18.0 + 12.0 * b + 6.0 * c) * x * x + (6.0 - 2.0 * b);
        return x / 6.0;
      }
      if (x < 2.0) {
        x = (-1.0 * b - 6.0 * c) * (x * x * x) + (6.0 * b + 30.0 * c) * x * x + (-12.0 * b - 48.0 * c) * x
            + (8.0 * b + 24.0 * c);
        return x / 6.0;
      }
      return 0.0;
    }
  }

  public static final class Quadratic extends Filter {
    public Quadratic() {
      super(false, 1.5, 1.0);
    }

    @Override
    public double f(double x) {
      if (x < 0)
        x = -x;
      if (x < 0.5)
        return 0.75 - x * x;
      if (x < 1.5) {
        x -= 1.5;
        return 0.5 * x * x;
      }
      return 0.0;
    }
  }

  public static final class Sinc extends Filter {
    public Sinc() {
      super(true, 4.0, 1.0);
    }

    @Override
    public double f(double x) {
      x *= Math.PI;
      if (x != 0.0)
        return Math.sin(x) / x;
      return 1.0;
    }
  }

  public static final class Triangle extends Filter {
    @Override
    public double f(double x) {
      if (x < 0.0)
        x = -x;
      if (x < 1.0)
        return 1.0 - x;
      return 0.0;
    }
  }

  /**
   * is this filter cardinal? ie, does func(x) = (x==0) for integer x?
   */
  final boolean cardinal;

  /** radius of nonzero portion */
  double support;

  /** blur factor (1=normal) */
  double blur;

  protected Filter() {
    this(true, 1.0, 1.0);
  }

  protected Filter(final boolean cardinal, final double support, final double blur) {
    this.cardinal = cardinal;
    this.support = support;
    this.blur = blur;
  }

  public double fWindowed(double x) {
    return x < -support || x > support ? 0 : f(x);
  }

  public abstract double f(double x);

  /**
   * Return the filter name.
   * 
   * @return the filter's name
   */
  public String getName() {
    return getClass().getSimpleName();
  }

  /**
   * @return the support
   */
  public double getSupport() {
    return support;
  }

  /**
   * @param support the support to set
   */
  public void setSupport(final double support) {
    this.support = support;
  }

  /**
   * @return the blur
   */
  public double getBlur() {
    return blur;
  }

  /**
   * @param blur the blur to set
   */
  public void setBlur(final double blur) {
    this.blur = blur;
  }
}
